### PR TITLE
fix(chart): Show metric label in tooltips

### PR DESCRIPTION
**Pull request by reynoldmorel**
### SUMMARY

This PR improves how metric names are displayed in chart tooltips by refining the column name matching logic for forecast series. The changes enhance forecast utility functions to better handle field name translation to human-readable labels, add clearer documentation for the verbose name resolution process, and introduce a lightweight test fixture to reduce test maintenance overhead.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

BEFORE

metric_label_bug

AFTER

https://github.com/user-attachments/assets/fc4a6032-1170-4034-af55-e37ef7e25c30

### TESTING INSTRUCTIONS

1. Create a dataset
2. Create a metric in the dataset
3. Add a label for the metric
4. Create a chart with the metric (line chart for example)
5. Add a time comparison for the chart
6. Make sure tooltips are enabled for the chart

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+  LegendType,
+  OrientationType,
+} from '../types';
+import { EchartsMixedTimeseriesProps } from './types';
+
+const mockedProps = {
+  // Chart dimensions
+  width: 787,
+  height: 190.1953125,
+
```
Comment (alexandrusoare) [inline superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts:24]: Why do we have these specific width and height?

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+      result_format: 'json',
+      applied_filters: [
+        {
+          column: 'time_start',
+        },
+      ],
+      rejected_filters: [],
+    },
+  ] as unknown as EchartsTimeseriesChartProps['queriesData'],
+
```
Comment (alexandrusoare) [inline superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: Is there a way to avoid the double type assignation?

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+      'time_start',
+      'time_start',
+    ];
+
+    legendData.sort((a, b) => a.localeCompare(b));
+    expectedResult.sort((a, b) => a.localeCompare(b));
+
+    expect(legendData).toEqual(expectedResult);
+  });
+});
```
Comment (alexandrusoare) [inline superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts:64]: I see that we have created a pretty complex object for one test. Is all the data necessary for this particular test? And can we also add more tests? We need to take into consideration negative scenarios as well

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+      'SUM(money_for_learning), 1 day ago',
+      'SUM(money_for_learning), 1 week ago',
+      'SUM(money_for_learning), 1 year ago',
+      'Testing count',
+      'Testing count, 1 day ago',
+      'Testing count, 1 week ago',
+      'Testing count, 1 year ago',
+      'time_start',
+    ]);
+  });
```
Comment (alexandrusoare) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts]: Ditto

File: superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts [hunk 5]
```diff
@@ -26,6 +26,41 @@ import {
...
+  const newVerboseMap: Record<string, string> = {};
+
+  Object.entries(label_map).forEach(([key, labels]) => {
+    if (labels) {
+      labels
+        .filter(l => verboseMap[l])
+        .forEach(l => {
+          const newKey = key.replaceAll(l, verboseMap[l]);
+          newVerboseMap[key] = newKey;
+        });
```
[the remark this answers is not in the document]
Comment (alexandrusoare) [reply superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts]: I think I agree with this

[the remark this answers is not in the document]
Comment (rebenitez1802) [reply superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts]: Yes this absolutely needed, please address this change request.

[the remark this answers is not in the document]
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts]: Addressed

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts (hunk 1 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts:24]: We don't need them I copied and past a model from the running app and created the mock from there.

Addressed by using AI to start removing unnecessary props for these tests.

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts (hunk 2 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: We would need to set all required objects, which would make the mock grow significantly

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts (hunk 3 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts:64]: There are actually no negative scenarios at all. The function I added `addLabelMapToVerboseMap`, on replaces a key when it exists if it doesn't then no replace operation is run. Which mean it will always return something and it is not supposed to fail, given the real scenarios, unless we change the data structure of `label_map` or `verboseMap` and the current test covers label replacements and non label replacements

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts (hunk 4 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts]: same as above

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+  rawFormData: {
+    datasource: '1__table',
+    viz_type: 'echarts_timeseries',
+    x_axis: 'ds',
+    time_compare: [],
+  } as unknown as EchartsTimeseriesChartProps['rawFormData'],
+
+  // Theme
+  theme: {},
+} as unknown as EchartsMixedTimeseriesProps;
```
Comment (geido) [inline superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: Do we need these `unknown`. Should we centralize the mocks? Probably out of scope for the work but something worth considering

File: superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts [hunk 7]
```diff
@@ -26,6 +26,147 @@ import {
...
+/**
+ * Replaces a label inside a compound key only if it appears as a complete
+ * word/token and contains at least one alphabetic character.
+ *
+ * @param key - The source string (typically a compound field name)
+ * @param label - The label to search for as a standalone token
+ * @param replacement - The human-readable value to replace the label with
+ * @returns The transformed key if a valid match exists, otherwise the original key
+ */
+const replaceLabelIfExists = (
```
Comment (geido) [inline superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts:48]: Needs a deeper review

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts (hunk 6 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: Addressed but needed to use `any`, `Partial` was causing other problems as well

File: superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.test.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      },
+    );
+
+    expect(result).toEqual({
+      testing_count: 'Testing count',
+      'testing_count, 1 day ago': 'Testing count, 1 day ago',
+      'testing_count, 1 week ago': 'Testing count, 1 week ago',
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (alexandrusoare) [reply superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.test.ts:59]: I agree can we add some more testing cases?

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import transformProps from './transformProps';
+import { mockedMixedTimeSeriesProps } from 'spec/fixtures/mockTimeSeries';
+
+describe('MixedTimeseries transformProps', () => {
```
Comment (EnxDev) [inline superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: I reccomand to follow this [guidelines ](https://superset.apache.org/developer_portal/testing/testing-guidelines) to use 'test' instead of `describe` with `it`

**Review by EnxDev — COMMENTED** [review]
Hey @reynoldmorel same thing here, what do you think about align this this [PR](https://github.com/apache/superset/pull/37396) as we may be adding logic that isn’t necessary if the data comes directly from the backend?

Feel free to follow thoose PRs #37396  and #37461 to keep patterns consistent, and please ensure the tests follow our best practices, thanks!

File: superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts (hunk 9 above)
Comment (reynoldmorel) [reply superset-frontend/plugins/plugin-chart-echarts/src/MixedTimeseries/transformProps.test.ts]: Addressed

**Comment by reynoldmorel** [comment]
> Hey @reynoldmorel same thing here, what do you think about align this this [PR](https://github.com/apache/superset/pull/37396) as we may be adding logic that isn’t necessary if the data comes directly from the backend?
> 
> Feel free to follow thoose PRs #37396 and #37461 to keep patterns consistent, and please ensure the tests follow our best practices, thanks!

Sure!! Will wait until the PR is merged to avoid any last change that could affect the implementation in the FE

File: superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts [hunk 10]
```diff
@@ -26,6 +26,147 @@ import {
...
+      const newLabelMap: Record<string, string> = labels
+        .filter(l => verboseMap[l])
+        .reduce(
+          (acc, label) => ({
+            ...acc,
+            [label]: verboseMap[label],
+          }),
+          {},
+        );
+
```
Comment (EnxDev) [inline superset-frontend/plugins/plugin-chart-echarts/src/utils/forecast.ts:143]: Would it make sense to define let newKey = key outside the inner forEach and accumulate changes there? This would avoid cases where a compound series key matches multiple labels and only the last replacement is applied.

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    'SUM(money_for_learning)',
+    'SUM(money_for_learning), 1 day ago',
+    'SUM(money_for_learning), 1 week ago',
+    'SUM(money_for_learning), 1 year ago',
+    'Testing count',
+    'Testing count, 1 day ago',
+    'Testing count, 1 week ago',
+    'Testing count, 1 year ago',
+    'time_start',
+  ];
```
[the remark this answers is not in the document]
Comment (EnxDev) [reply superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/transformProps.test.ts:47]: @reynoldmorel  what do you think avout that?

**Comment by EnxDev** [comment]
Hey @reynoldmorel, are you planning to move forward with this? Let me know if you’d like a hand; happy to help!

**Comment by rusackas** [comment]
@reynoldmorel thanks for this — the verbose-map approach looks reasonable and there's a test. Just updated the branch to (hopefully) pass CI... then we can take it from there, but thus far it LGTM.

**Comment by EnxDev** [comment]
## EnxDev's Review Agent — apache/superset#37229 · HEAD e6e5600
**comment** — Clean, well-tested fix for compound-key tooltip labels. One latent crash on missing `label_map` and a multi-word-label gap before LGTM.

### 🔴 Functional
- **`utils/forecast.ts:91` + `MixedTimeseries/transformProps.ts:150-159`** — `addLabelMapToVerboseMap`'s first param has no default, and `labelMap`/`labelMapB` are destructured without defaults (`transformProps.ts:150,152`) and passed straight in. `Object.entries(label_map)` (`forecast.ts:132`) throws `Cannot convert undefined or null to object` if a query omits `label_map`. The Timeseries path guards this (`label_map = {}`, `Timeseries/transformProps.ts:211`) and existing Mixed code already optional-chains `labelMap?.[seriesName]` (`transformProps.ts:461`) — so the asymmetry is real. Default the helper's first param to `{}`? (test: `transformProps`/`addLabelMapToVerboseMap` with `label_map` undefined)

### 🟡 Should-fix
- **`utils/forecast.ts:144`** — `key.split(' ')` tokenizes on spaces, so any `verboseMap` key containing a space (quoted identifier, custom metric label like `Order Total`) never matches and the compound series keeps its raw label. Single-token keys work; multi-word labels silently don't. Run the replacement over the full `key` (or tokenize on `, `) instead of single spaces.
- **`utils/forecast.test.ts`** — no case covers the missing/empty `label_map` path — the exact crash in the Functional item. Add one; it's red before the fix above, green after.

### 🔵 Nits
- `utils/forecast.ts:41,60` — `replaceLabelIfExists` JSDoc claims it requires "at least one alphabetic character" and ignores numeric-only matches, but the body (`forecast.ts:78-80`) has no such check; the `"12"` vs `"123"` example only holds via the word boundaries. Align the doc with the code.
- `spec/fixtures/mockTimeSeries.ts:162,269` — blanket `as any` violates the no-`any` rule. Type the fixtures against `EchartsTimeseriesChartProps`/`EchartsMixedTimeseriesProps` (or use narrow casts on the few loose fields) rather than casting the whole object.

### 🙌 Praise
- `utils/forecast.test.ts` — the `(?<!\w)…(?!\w)` boundary matching cleanly avoids substring collisions (`account` not matching inside `testing_account`, `count` not inside `account`), and the test cases prove it. Nice.


_Reviewed by EnxDev's Review Agent — @EnxDev · HEAD e6e5600._

**Comment by rusackas** [comment]
@reynoldmorel one thing before this can land: `addLabelMapToVerboseMap`'s `label_map` param has no default, so `Object.entries(label_map)` throws if a query context omits `label_map` - can you default it to `{}` like the existing Timeseries fallback does?

**Comment by reynoldmorel** [comment]
> @reynoldmorel one thing before this can land: `addLabelMapToVerboseMap`'s `label_map` param has no default, so `Object.entries(label_map)` throws if a query context omits `label_map` - can you default it to `{}` like the existing Timeseries fallback does?

Hi!

Sorry I couldn't see any of these messages before, is it possible to have empty label_map being returned by business logic?

**Comment by rusackas** [comment]
@reynoldmorel label_map is always in the chart-data payload as a dict (worst case {}) per query_context_processor.py, so defaulting it to {} in the frontend is safe, I think. Can you add that and rebase off master?

**Comment by reynoldmorel** [comment]
@rusackas - Sure I will try to do it between today and tomorrow!

**Comment by rusackas** [comment]
@reynoldmorel following up! The `label_map` default fix from last month never landed, so `addLabelMapToVerboseMap` will still throw if a query omits it. Also still needs a rebase, and there's a pile of bot threads on the test expectations to clear. Ping me when it's ready for another look. Say the word if you want a hand here.

**Comment by rusackas** [comment]
I'll rebase this off `master` since maintainer edits are enabled, the tooltip label fix still looks worth landing. Worth checking Bito's suggestions on `forecast.ts` before it's ready though.
